<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [ ]:
import pandas as pd
import psycopg2
import warnings
warnings.filterwarnings('ignore', message = '.*SQLAlchemy connectable')

In [ ]:
DBNAME = '****'
USER = '****'
PASSWORD = '****'
HOST = '****'
PORT = 5432

In [ ]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [ ]:
query_3_1 = f'''select * /*выбираем все столбцы из таблицы vacancies*/
    from public.vacancies
'''

In [ ]:
# используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_3_1 = pd.read_sql_query(query_3_1, connection)

print('количество вакансий=', df_3_1['id'].count()) # считаем количество строк в таблице

display(df_3_1) # выводим таблицу на экран

df_3_1.info() # информация о таблице


количество вакансий= 49197


id                                               name  \
0      55312386                                Компьютерный Мастер   
1      55843014                            Системный администратор   
2      54525964                      Lead Java Developer to Poland   
3      54525965                      Lead Java Developer to Poland   
4      55354053  Специалист службы поддержки с техническими зна...   
...         ...                                                ...   
49192  55589514           Специалист по информационным технологиям   
49193  55029668                     Руководитель процессного офиса   
49194  66172263                         Начальник сервисной службы   
49195  66180086                             Senior IOS разработчик   
49196  55547842                                   DataOps Engineer   

                                              key_skills          schedule  \
0      Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1      Средства криптографической защиты информации\t...       Полный день   
2        Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3        Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                                   None  Удаленная работа   
...                                                  ...               ...   
49192  Электронный документооборот\t1С: Предприятие 8...       Полный день   
49193  Оптимизация бизнес-процессов\tСтратегическое м...       Полный день   
49194  Организаторские навыки\tУправление персоналом\...       Полный день   
49195  Swift\tRealm\tiOS\tАнглийский язык\tMVP\tMVVM\...       Полный день   
49196                                               None  Удаленная работа   

               experience           employment  salary_from  salary_to  \
0               Нет опыта     Полная занятость      64000.0        NaN   
1           От 3 до 6 лет     Полная занятость          NaN        NaN   
2           От 3 до 6 лет     Полная занятость          NaN        NaN   
3           От 3 до 6 лет     Полная занятость          NaN        NaN   
4               Нет опыта  Частичная занятость      15000.0        NaN   
...                   ...                  ...          ...        ...   
49192       От 3 до 6 лет     Полная занятость          NaN        NaN   
49193       От 3 до 6 лет     Полная занятость          NaN        NaN   
49194  От 1 года до 3 лет     Полная занятость      60000.0        NaN   
49195  От 1 года до 3 лет     Полная занятость     146456.0   244094.0   
49196       От 3 до 6 лет     Полная занятость          NaN        NaN   

       area_id  employer_id  
0         1516      5724811  
1          130      4903713  
2          160        69961  
3          159        69961  
4         1955         1740  
...        ...          ...  
49192       66      5109332  
49193       26        51333  
49194     1930      4094700  
49195      160      4469939  
49196      160      4978627  

[49197 rows x 10 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49197 entries, 0 to 49196
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           49197 non-null  int64  
 1   name         49197 non-null  object 
 2   key_skills   40741 non-null  object 
 3   schedule     49197 non-null  object 
 4   experience   49197 non-null  object 
 5   employment   49197 non-null  object 
 6   salary_from  21485 non-null  float64
 7   salary_to    12690 non-null  float64
 8   area_id      49197 non-null  int64  
 9   employer_id  49197 non-null  int64  
dtypes: float64(2), int64(3), object(5)
memory usage: 3.8+ MB


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [ ]:
query_3_2 = f'''select * /*выбираем все столбцы из таблицы employers*/
    from public.employers
'''

In [ ]:
df_3_2 = pd.read_sql_query(query_3_2, connection)

print('количество работодателей=', df_3_2['id'].count()) # считаем количество строк в таблице

display(df_3_2) # выводим таблицу на экран


количество работодателей= 23501


id                                               name  area
0         2393                   Программный Продукт, ИТ-компания     1
1        72977                                          БАРС Груп    88
2         3155  Бест, Торгово-производственная компания, Екате...     3
3          675                                   КОРУС Консалтинг     2
4      1840010  филиал ФКУ Налог-Сервис ФНС России в Республик...    88
...        ...                                                ...   ...
23496  1441774                                              МарКо    88
23497  4718575                                      MKomov Studio    73
23498  3702770                                  АКИБ Ипотека-банк  2759
23499  5636092                                        ГИМ-КЛИНИКА    99
23500  2064382                                             Рыббла   104

[23501 rows x 3 columns]

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [ ]:
query_3_3 = f'''select *
    from public.areas
'''

In [ ]:
df_3_3 = pd.read_sql_query(query_3_3, connection)

print('количество регионов=', df_3_3['id'].count()) # считаем количество строк в таблице

display(df_3_3) # выводим таблицу на экран

количество регионов= 1362


id              name
0     2758           Тбилиси
1        8            Майкоп
2     1180          Нерюнгри
3     1240       Новокузнецк
4        2   Санкт-Петербург
...    ...               ...
1357  5043  Звёздный Городок
1358  3934       Воротынск\n
1359  4825             Чемал
1360  3708         Усть-Нера
1361  2432         Каневская

[1362 rows x 2 columns]

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [ ]:
query_3_4 = f'''select *
    from public.industries
'''

In [ ]:
df_3_4 = pd.read_sql_query(query_3_4, connection)

print('количество сфер деятельности=', df_3_4['id'].count())

display(df_3_4)

***

## По предварительному анализу данных можно сделать следующие выводы:
1. Так как в России на [1 октября 2023г. проживало 80 286 415 ](https://bdex.ru/naselenie/russia/?ysclid=lnhc3oqwp0407135374) трудоспособных человек получается, что на каждые 10 000 человек 6 вакансий по компьютерным технологиям. А на [11 безработных ]( https://lenta.ru/news/2023/04/08/golikova/) получается одна вакансия по компьютерным технологиям. Что говорит о не хватке специалистов по компьютерным технологиям.
2. Количество работодателей по данным из [HeadHunter](https://ru.wikipedia.org/wiki/HeadHunter) свыше 515 тыс. компаний из них ищут сотрудников, а по компьютерным технологиям 23.5 тыс. компаний то есть 4.5% что довольно много и рост продолжается. Это говорит о том, что все больше компаний начинают применять компьютерные технологии у себя на производстве.  
3. [Количество городов в России 1119](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8), а в нашей таблицы  1362 региона это означает, что практически в любом городе  найдутся вакансии по компьютерным технологиям. Но в почти половине регионах вакансии отсутствуют (см. ниже). По этому в регионах с низкой протностью населения вакансий мало. 
4. Основных [сфер деятельности 13:](https://profstandart-rosmintrud.ru/kakie-sfery-deyatelnosti-sushhestvuyut/?ysclid=lnhev2pcbz87396995#%D0%A2%D0%B5%D1%85%D0%BD%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%B8%D0%B5) Технические, Экономические, Творческие, Военные профессии, Профессии связанные с путешествиями, Медицинские, Педагогические, Юридические, Работа с людьми, Безопасность, Аграрные и животноводческие, Транспортные профессии, Лингвистические. В нашей таблице 294 сферы деятельности это означает, что компьютерные технологии проникли практически во все эти сферы тем или иным образом 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [ ]:
query_4_1 = f'''select 
    a.name area,                        /* выводим наименование регионов с алиусом area */
    count(v.id) cnt                     /* выводим сумму не пустых значений в каждом регионе */
    
    from public.vacancies v left join    /* обединяем таблицы vacancies и area с алиасами 'v' и 'a' */
        public.areas a on v.area_id=a.id /* соответственно при помоши left join по v.area_id и a.id */
        
    group by 1                           /* груперуем таблицу по первому столбцу a.name */
    order by 2 desc                      /* сортируем по убыванию второго столбца */
            
'''

In [ ]:
df_4_1= pd.read_sql_query(query_4_1, connection)
df_4_1

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
query_4_2 = f'''select 
    count(id) cnt                 /* считаем количество не пустых строк */
            
    from public.vacancies         /* подключаем таблицу vacancies */
            
    /* записываем условие чтобы хтябы одно из значений таблицы было заполнено */
     where (salary_to is not null) or (salary_from is not null)
'''

In [ ]:
df_4_2 = pd.read_sql_query(query_4_2, connection)
print("количество вакансий с уазанной зарплаттой= ", df_4_2.loc[0,'cnt'])

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [ ]:
query_4_3 = f'''select 
    avg(salary_to) avg_to , /* вычисляем среднее значение верхней границы зарплатной вилки */
    avg(salary_from) avg_from /* вычисляем среднее значение нижней границы зарплатной вилки */
            
    from public.vacancies  /* подключаем таблицу vacancies */
            
    /* записываем условие чтобы хтябы одно из значений таблицы было заполнено */
    where (salary_to is not null) or (salary_from is not null)
'''

In [ ]:
df_4_3= pd.read_sql_query(query_4_3, connection)
display(round(df_4_3))

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
query_4_4 = f'''select  
    distinct /* выбираем уникальные значения */

    v.schedule "типа рабочего графика", /* все соетания рабочего гафика и типа трудоустройства */
    v.employment "типа трудоустройства",
    
    count(*) "количество вакансий"  /* количество уникальных значений */
    
    from public.vacancies v /* подключаем таблицу vacancies */
    
    group by 1,2 /* груируем по первому а затем по второму столбцу */
    order by 3 desc /* сортируем по убыванию третьего столбца */
'''

In [ ]:
df_4_4= pd.read_sql_query(query_4_4, connection)
display(df_4_4)

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [ ]:
query_4_5 = f'''select  
    distinct                             /* выбираем уникальные значения */
    v.experience "Требуемый опыт работы",/* выбираем столбец experience (опыт работы)*/
    count(*) cnt                         /* количество уникальных значений */
    
    from public.vacancies v              /* подключаем таблицу vacancies */
    
    group by 1                           /* груируем по первому  столбцу */
    order by 2                           /* сортируем по возрастанию вторго столбца */
'''

In [ ]:
df_4_5= pd.read_sql_query(query_4_5, connection)
display(df_4_5)

***

## По анализу вакансий можно сделать следующие выводы:
1.	Если рассматривать вакансии по территориальному признаку, то наибольшее количество вакансий приходится на города миллионщики «Москва, Новосибирск и т.д.», где плотность населения выше. Но и в остальных регионах есть или были вакансии. 
2.	Вакансий с установленной заработной платой примерно половина от общего числа и средняя от 71 тыс. руб. до 110.5 тыс. руб., у остальных вакансий зарплата либо договорная, либо озвучивается в тексте объявления. 
3.	Работодатели предпочитают искать работников на полный рабочий день и полную занятость 35.3 тыс. вакансий, также удаленная работа и полная занятость 7.8 тыс. вакансии. Остальные сочетания менее привлекательны работодателям самая не привлекательная пара сменный график проектная работа одна вакансия
4.	В связи с тем, что у нас наблюдается нехватка специалистов по компьютерным технологиям многие компании соглашаются брать специалистов с минимальным опытом от 1 года до 3 лет и таких компаний больше половины 26.2 тыс. Четверть компаний берут специалистов с опытом от 3 до 6 лет 14.5 тыс.  Без опыта работы тоже требуются 7.2 тыс. и немного компаний 1.3 тыс. хотят иметь специалиста с опытом работы более 6 лет.   

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [ ]:
query_5_1 = f'''select  distinct   /* выбираем уникальные значения */
                e.name "работодатели",
                count(v.id) "количествo вакансий"
            from public.vacancies v full join public.employers e on e.id= v.employer_id
            group by 1
            order by 2 desc
'''

In [ ]:
df_5_1= pd.read_sql_query(query_5_1, connection)
display(df_5_1.head(10))

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
query_5_2 = f'''select a.name регион,
       count(distinct e.id) количествo_работодателей,
       count(distinct v.id) количествo_вакансий
from areas a
     left join employers e on a.id = e.area
     left join vacancies v on v.area_id = a.id
where (v.id is null)
group by a.id

order by 2 desc  
'''

In [ ]:
df_5_2= pd.read_sql_query(query_5_2, connection)
df_5_2

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [ ]:
query_5_3 = f'''select 
     e.name название_работодателя,
     count(distinct v.area_id) количествo_регионов
       
from  employers e  /*обьединяем таблици работадателей и вакансий */
     left join vacancies v on v.employer_id = e.id
/*where v.area_id is null*/
group by 1 /*групмоуем по наменовнию работадателей*/
order by 2 desc /* сортируем по уменьшению количества регионов*/
'''

In [ ]:
df_5_3= pd.read_sql_query(query_5_3, connection)
df_5_3

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [ ]:
query_5_4 = f'''select 
    e.name название_работодателя
       
from employers e left join
    EMPLOYERS_INDUSTRIES ei on e.id = ei.employer_id
where ei.industry_id is null

'''

In [ ]:
df_5_4= pd.read_sql_query(query_5_4, connection)
df_5_4

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [ ]:
query_5_5 = f'''select 
    e.name
from employers e left join
    EMPLOYERS_INDUSTRIES ei on e.id = ei.employer_id
group by 1
having count(ei.industry_id)=4
order by 1 
/*offset 2 limit 1*/
'''

In [ ]:
df_5_5= pd.read_sql_query(query_5_5, connection)
df_5_5

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [ ]:
query_5_6 = f'''select 
   count(*)
from employers e left join
    EMPLOYERS_INDUSTRIES ei on e.id = ei.employer_id 
    left join industries i on ei.industry_id=i.id
where i.name like 'Разработка программного обеспечения'
''' 

In [ ]:
df_5_6= pd.read_sql_query(query_5_6, connection)
df_5_6

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [ ]:
from urllib.parse import quote_plus
url = quote_plus('https://ru.wikipedia.org/wiki/Города-миллионеры_России', safe=':/')
tables = pd.read_html(url)
mln_cities = tables[1]['Город'].values.tolist()
mln_cities

In [ ]:
query_5_7 = f'''select  
    a.name "регион",
    count(*) "количество вакансий"
from vacancies v left join  EMPLOYERS e on v.employer_id=e.id /* обединяем три таблицы вакансии регионы работадатели*/
    left join areas a on a.id=v.area_id
where (a.name in  ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград'))and e.name like 'Яндекс'
group by a.name
union all
select
    'total',
    count(*)
from vacancies v left join  EMPLOYERS e on v.employer_id=e.id
    left join areas a on a.id=v.area_id
where (a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград'))and e.name like 'Яндекс'
order by 2 desc
''' 

In [ ]:
df_5_7= pd.read_sql_query(query_5_7, connection)
df_5_7

***

## Анализ работодателей:
1.	По количеству вакансий на первые места попадают крупные компании такие как:
*	Яндекс
*	Ростелеком
*	Тинькофф
*	СБЕР
*	И т.д.  
У каждой компаний более 400 вакансий и сами компании встречаются по всей стране, найти в этих компаниях работу проще. По регионам компании распределены не равномерно в 593 регионах нет вакансий, а это возможно потому что работодатели нашли работников и пока у них нет вакансий. Поэтому чтобы устроится работать в этих регионах надо либо переезжать в другой город, либо работать по удаленьки, либо работать вахтовым методом.  
2.	У крупных кампаний распространение на большее количество регионов, например:
*	Яндекс 181 регион
*	Ростелеком 152 региона
*	Спецремонт 116 регионов
*	И т.д.

3.	Многие компании занимаются разной деятельностью даже не в IT сфере у некоторых несколько сфер деятельности, а некоторые вообще не указывают свою сферу деятельности конечно можно определить из названия компании, но не всегда на пример **Нозиков Денис Михайлович**. Достаточное количество компаний занимаются разработкой программного обеспечения 3553 компании куда требуются специалисты 
4.	Возьмём, например, Яндекс ему требуются много специалистов и во всех городах миллионщиках нужны специалисты от 21 до 53 человек и вообще у него 1933 вакансии по всей стране. Яндекс набирает много людей, что в последствии может его еще выше поднять в IT сфере. 

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
query_6_1 = f'''select  
    count(*) cnt /*считаем количество вакансий*/
from vacancies
where lower(name) like '%data%' or lower(name) like '%данн%'/*выделяем вакансии в названии которых есть
                        сочетание 'data' и 'данн'*/
'''

In [ ]:
df_6_1= pd.read_sql_query(query_6_1, connection)
df_6_1

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
query_6_2 = f'''select  
    count(name) cnt /*считаем количество вакансий*/
from vacancies

/*выделяем вакансии в названии которых есть сочетание 'data scientist', 'data science', 
'исследователь данных', 'machine learning', 'машинн%обучен' и 'ML' без 'HTML */
where ((lower(name) like '%data scientist%') 
    or (lower(name) like '%data science%') 
    or (lower(name) like '%исследователь данных%') 
    or (lower(name) like '%machine learning%')
    or ((lower(name) like '%ml%') and (lower(name)  not like '%html%'))
    or (lower(name) like '%машинн%обучен%'))
    
/* выделяем вакансии для junior*/    
    and((lower(name) like '%junior%')
    or (experience like '%Нет опыта%' )
    or (employment like '%Стажировка%')) 
    '''

In [ ]:
df_6_2= pd.read_sql_query(query_6_2, connection)
df_6_2

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
query_6_3 = f'''select  
    count(name) cnt /*считаем количество вакансий*/
from vacancies
/* выделяем вакансии DS*/
where ((lower(name) like '%data scientist%') 
    or (lower(name) like '%data science%') 
    or (lower(name) like '%исследователь данных%') 
    or (lower(name) like '%machine learning%')
    or ((name like '%ML%') and (lower(name)  not like '%html%'))
    or (lower(name) like '%машинн%обучен%'))
    
    /* выделяем вакансии для SQL и postgres*/ 
    and((key_skills like '%SQL%') or (lower(key_skills) like '%postgres%'))
    '''

In [ ]:
df_6_3= pd.read_sql_query(query_6_3, connection)
df_6_3

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
query_6_4 = f'''select  
   
    count(name) cnt
from vacancies
/* выделяем вакансии DS*/
where ((lower(name) like '%data scientist%') 
    or (lower(name) like '%data science%') 
    or (lower(name) like '%исследователь данных%') 
    or (lower(name) like '%machine learning%')
    or ((name like '%ML%') and (lower(name)  not like '%html%'))
    or (lower(name) like '%машинн%обучен%'))
    
    /* выделяем вакансии для Python */ 
    and((key_skills like '%Python%'))
'''  

In [ ]:
df_6_4= pd.read_sql_query(query_6_4, connection)
df_6_4

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
query_6_5 = f'''select  
    /* вычесляем среднее количество навыков для вакансий DC*/
    sum(length(key_skills) - length(REPLACE(key_skills, CHR(9), ''))+1)::real / count(key_skills) "количество навыков"
from vacancies
/* выделяем вакансии DS*/
where ((lower(name) like '%data scientist%') 
    or (lower(name) like '%data science%') 
    or (lower(name) like '%исследователь данных%') 
    or (lower(name) like '%machine learning%')
    or ((name like '%ML%') and (lower(name)  not like '%html%'))
    or (lower(name) like '%машинн%обучен%'))
    
'''  

In [ ]:
df_6_5= pd.read_sql_query(query_6_5, connection)
print(df_6_5.round(2))

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
query_6_6 = f'''with sum1 as (select /* создаем временную таблицу sum1 */
    
    experience , 
    (salary_from + salary_to)/2 ЗП  /* вычисляем среднюю ЗП если даны верхняя и нижняя граница ЗП*/
   
from vacancies
/* выделяем вакансии DS*/
where ((lower(name) like '%data scientist%') 
    or (lower(name) like '%data science%') 
    or (lower(name) like '%исследователь данных%') 
    or (lower(name) like '%machine learning%')
    or ((name like '%ML%') and (lower(name)  not like '%html%'))
    or (lower(name) like '%машинн%обучен%'))
    
    /* выбираем не нуливые данные ЗП верхней и нижней границы одновременно */
    and((salary_from is not null) and (salary_to is not null))
 
union all
select  
    experience,
    coalesce(salary_from, 0) +coalesce(salary_to, 0) /* вычисляем с ЗП если даны верхняя или нижняя граница ЗП*/

from vacancies
/* выделяем вакансии DS*/
where ((lower(name) like '%data scientist%') 
    or (lower(name) like '%data science%') 
    or (lower(name) like '%исследователь данных%') 
    or (lower(name) like '%machine learning%')
    or ((name like '%ML%') and (lower(name)  not like '%html%'))
    or (lower(name) like '%машинн%обучен%'))
    
    /* выбираем не нуливые данные ЗП отдельно верхней или нижней границы  */
    and ((salary_from is not null) or (salary_to is not null))
    and not ((salary_from is not null) and (salary_to is not null))

 ) 
select
experience "опыт работы",
avg(ЗП) ЗП  /*вычисляем среднюю ЗП в зависемости от опыта работы */
from sum1 
group by 1
'''

In [ ]:
df_6_6= pd.read_sql_query(query_6_6, connection)
print(df_6_6.round())

***

## Анализ вакансий которые имеют отношение к данным
1.	Вакансий, относящихся к данным, насчитывается 1771 шт. это 3.5% от общего количества вакансий, что довольно много, но это очень мало по сравнению с численностью работоспособного населения.
2.	 Для начинающих дата-сайентистов насчитывается 51 вакансия из 536, а это 10%, что означает нехватку специалистов, хотя это 2 – 3 выпуска любой школы IT 
3.	Знания SQL и Python являются одним из ключевых навыков для работы с DS  201 и 351 вакансия требует их наличия.
4.	Знание SQL и Python не являются единственными навыками, которые требует работодатель в среднем требуется более 6 навыков (6.4).
5.	ЗП очень зависит от опыта работы так без опыта ЗП около 75 тыс. руб. от 1 года до 3 лет 140 тыс. руб.  от 3 до 6 лет 240 тыс. руб. ЗП довольно высокие даже для начинающих специалистов 

# Общий вывод по проекту

## ВЫВОДЫ
1.	Вакансий в IT сфере довольно много и IT специалисты требуются наверно во всех сферах деятельности людей 
2.	В основном IT специалисты требуются в крупных городах миллионщиках с малым опытом работы и ЗП около 100 тыс. руб. что говорит о нехватке специалистов в больших городах.
3.	Основными работодателями являются большие компании, которые распространены по всей стране и работают в сфере финансов или информационных услуг. Так же много средних компаний, которым нужен один два специалиста. 
4.	Вакансий, которые имеют отношение к данным не так уж и много и требования, которые предъявляются к специалистам довольно высокие нужно иметь 6 и более навыков чтобы устроится на работу и средняя зарплата у них выше.